File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/level/block/entities/BallLootBlockEntity.java
package generations.gg.generations.core.generationscore.common.world.level.block.entities;

import generations.gg.generations.core.generationscore.common.GenerationsCore;
import generations.gg.generations.core.generationscore.common.util.GenerationsUtils;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.core.HolderLookup;
import net.minecraft.core.NonNullList;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.ListTag;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.block.HorizontalDirectionalBlock;
import net.minecraft.world.level.block.state.BlockState;

import java.time.Instant;
import java.util.*;
import java.util.stream.Collectors;

public class BallLootBlockEntity extends ModelProvidingBlockEntity {
    private UUID owner = null;
    private final List<LootClaim> claims = new ArrayList<>();
    private LootMode lootMode = LootMode.ONCE_PER_PLAYER;
    private boolean doesCustomDrops = false;
    private boolean visible = true;
    private NonNullList<ItemStack> customDrops = null;

    @Override
    protected void readNbt(CompoundTag nbt, HolderLookup.Provider provider) {
        owner = nbt.contains("owner") ? nbt.getUUID("owner") : null;

        claims.clear();
        if(nbt.contains("claims")) {
            var listTag = nbt.getList("claims", CompoundTag.TAG_COMPOUND);

            for (int i = 0; i < listTag.size(); i++) {
                var tag = listTag.getCompound(i);
                var uuid = tag.getUUID("uuid");
                var instant = tag.getLongArray("instant");

                claims.add(new LootClaim(uuid, Instant.ofEpochSecond(instant[0], instant[1])));
            }
        }

        lootMode = LootMode.valueOf(nbt.getString("lootMode"));
        doesCustomDrops = nbt.getBoolean("doesCustomDrops");
        visible = nbt.getBoolean("visible");

        if(nbt.contains("customDrops"))
            customDrops = nbt.getList("customDrops", CompoundTag.TAG_COMPOUND).stream().map(CompoundTag.class::cast).map(ItemStack::of).collect(Collectors.toCollection(NonNullList::create));
    }

    @Override
    protected void writeNbt(CompoundTag nbt, HolderLookup.Provider provider) {
        if(owner != null)nbt.putUUID("owner", owner);
        nbt.put("claims", claims.stream().map(a -> {
            var tag = new CompoundTag();
            tag.putUUID("uuid", a.uuid);
            var instant = a.time;
            tag.putLongArray("instant", new long[] { instant.getEpochSecond(), instant.getNano() });

            return tag;
        }).collect(ListTag::new, AbstractList::add, AbstractCollection::addAll));

        nbt.putString("lootMode", lootMode.name());
        nbt.putBoolean("doesCustomDrops", doesCustomDrops);
        nbt.putBoolean("visible", visible);

        if(customDrops != null)
            nbt.put("customDrops", customDrops.stream().map(GenerationsUtils::toCompoundTag).collect(ListTag::new, AbstractList::add, AbstractCollection::addAll));
    }

    public BallLootBlockEntity(BlockPos pPos, BlockState pBlockState) {
        super(GenerationsBlockEntities.BALL_LOOT.get(), pPos, pBlockState);
    }

//    @Override
//    public String getVariant() {
//        return getBlockState().getBlock() instanceof BallLootBlock loot ? loot.getType() : "poke";
//    }

    public UUID getOwner() {
        return owner;
    }

    public void setOwner(UUID owner) {
        this.owner = owner;
    }

    public boolean canClaim(UUID playerUUID) {
        if (lootMode == LootMode.UNLIMITED) return true;
        
        Optional<LootClaim> claim = this.getLootClaim(playerUUID);

        if (claim.isEmpty()) {
            addClaimer(playerUUID);
            return true;
        }
        if (this.lootMode.isTimeEnabled())
            if (claim.get().isExpired()) {
                claim.get().refresh();
                return true;
            }
        return false;
    }

    public Optional<LootClaim> getLootClaim(UUID playerUUID) {
        return this.claims.stream().filter(claim -> claim.uuid.equals(playerUUID)).findFirst();
    }

    public void addClaimer(UUID playerUUID) {
        this.claims.add(new LootClaim(playerUUID, Instant.now().plus(GenerationsCore.CONFIG.lootTime)));
    }


    public void removeClaimer(UUID playerUUID) {
        this.claims.removeIf(a -> a.uuid.equals(playerUUID));
    }

    public boolean shouldBreakBlock() {
        return lootMode.isBreakable();
    }

    public void setLootMode(LootMode lootMode) {
        this.lootMode = lootMode;
    }

    public LootMode getLootMode() {
        return this.lootMode;
    }

    public boolean isCustomDrop() {
        return this.doesCustomDrops && this.customDrops != null;
    }

    public NonNullList<ItemStack> getCustomDrops() {
        return this.customDrops;
    }

    public void setCustomDrops(ItemStack ... customDrops) {
        this.doesCustomDrops = true;
        this.customDrops = NonNullList.of(ItemStack.EMPTY, customDrops);
    }

    public void addCustomDrop(ItemStack drop) {
        this.doesCustomDrops = true;
        if(this.customDrops == null)
            this.customDrops = NonNullList.create();
        this.customDrops.add(drop);
    }

    public boolean isVisible() {
        return visible;
    }

    public void setVisible(boolean visible) {
        this.visible = visible;
        sync();
    }

    public enum LootMode {
        ONCE_PER_PLAYER, TIMED, ONCE, UNLIMITED;

        public boolean isDropOnce() {
            return this == ONCE_PER_PLAYER || this == ONCE;
        }

        public boolean isTimeEnabled() {
            return this == TIMED;
        }

        public boolean isBreakable() {
            return this == ONCE;
        }
    }

    @Override
    public float getAngle() {
        if (getBlockState().hasProperty(HorizontalDirectionalBlock.FACING)) {
            var dir = getBlockState().getValue(HorizontalDirectionalBlock.FACING);

            switch (dir) {
                case EAST -> dir = Direction.WEST;
                case WEST -> dir = Direction.EAST;
                default -> {}
            }

            return dir.getOpposite().toYRot();
        }
        return 0f;
    }
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/level/block/entities/BallLootBlockEntity.kt
package generations.gg.generations.core.generationscore.common.world.level.block.entities

import com.mojang.serialization.Codec
import com.mojang.serialization.DataResult
import com.mojang.serialization.DynamicOps
import com.mojang.serialization.JsonOps
import generations.gg.generations.core.generationscore.common.GenerationsCore
import generations.gg.generations.core.generationscore.common.client.render.rarecandy.instanceOrNull
import generations.gg.generations.core.generationscore.common.util.GenerationsUtils
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.core.HolderLookup
import net.minecraft.core.NonNullList
import net.minecraft.nbt.CompoundTag
import net.minecraft.nbt.ListTag
import net.minecraft.nbt.NbtOps
import net.minecraft.nbt.Tag
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.block.HorizontalDirectionalBlock
import net.minecraft.world.level.block.state.BlockState
import java.time.Instant
import java.util.*
import java.util.function.BiConsumer
import java.util.function.Function
import java.util.function.Supplier
import java.util.stream.Collectors

private fun <A, B> Codec<A>.write(value: A, ops: DynamicOps<B>): B? {
    return this.encodeStart(ops, value).result().orElse(null)
}

private fun <A> Codec<A>.write(value: A): Tag? {
    return write(value, NbtOps.INSTANCE)
}

private fun <A, B> Codec<A>.read(value: B, ops: DynamicOps<B>): A? {
    return this.parse(ops, value).result().orElse(null)
}

class BallLootBlockEntity(pPos: BlockPos, pBlockState: BlockState) :
    ModelProvidingBlockEntity(GenerationsBlockEntities.BALL_LOOT.get(), pPos, pBlockState) {
    //    @Override
    //    public String getVariant() {
    //        return getBlockState().getBlock() instanceof BallLootBlock loot ? loot.getType() : "poke";
    //    }
    var owner: UUID? = null
    private val claims: MutableList<LootClaim> = ArrayList()
    var lootMode: LootMode = LootMode.ONCE_PER_PLAYER
    private var doesCustomDrops = false
    private var _visible = true

    var visible: Boolean
        get() = _visible
        set(value) {
            _visible = value
            sync()
        }

    var customDrops: NonNullList<ItemStack>? = null
        private set

    override fun readNbt(nbt: CompoundTag, provider: HolderLookup.Provider) {
        owner = if (nbt.contains("owner")) nbt.getUUID("owner") else null

        claims.clear()
        if (nbt.contains("claims")) {
            val listTag = nbt.getList("claims", CompoundTag.TAG_COMPOUND.toInt())

            for (i in listTag.indices) {
                val tag = listTag.getCompound(i)
                val uuid = tag.getUUID("uuid")
                val instant = tag.getLongArray("instant")

                claims.add(LootClaim(uuid, Instant.ofEpochSecond(instant[0], instant[1])))
            }
        }

        lootMode = LootMode.valueOf(nbt.getString("lootMode"))
        doesCustomDrops = nbt.getBoolean("doesCustomDrops")
        visible = nbt.getBoolean("visible")

        if (nbt.contains("customDrops")) customDrops =
            nbt.getList("customDrops", CompoundTag.TAG_COMPOUND.toInt()).stream()
                .map { obj: Tag? -> CompoundTag::class.java.cast(obj) }.map { a: CompoundTag? ->
                    ItemStack.CODEC.parse(
                        NbtOps.INSTANCE, a
                    ).mapOrElse(
                        Function.identity()
                    ) { error: DataResult.Error<ItemStack?>? -> ItemStack.EMPTY }
                }.collect(
                    Collectors.toCollection(
                        Supplier { NonNullList.create() })
                )
    }

    override fun writeNbt(nbt: CompoundTag, provider: HolderLookup.Provider) {
        owner.ifNotNull { nbt.putUUID("owner", it) }
        nbt.put(
            "claims", claims.stream().map<CompoundTag> { a: LootClaim ->
                val tag = CompoundTag()
                tag.putUUID("uuid", a.uuid)
                val instant = a.time
                tag.putLongArray("instant", longArrayOf(instant.epochSecond, instant.nano.toLong()))
                tag
            }.collect(
                Supplier<ListTag> { ListTag() },
                BiConsumer<ListTag, CompoundTag> { obj: ListTag, e: CompoundTag -> obj.add(e) },
                BiConsumer<ListTag, ListTag> { obj: ListTag, c: ListTag -> obj.addAll(c) })
        )

        nbt.putString("lootMode", lootMode.name)
        nbt.putBoolean("doesCustomDrops", doesCustomDrops)
        nbt.putBoolean("visible", visible)

        customDrops?.map(ItemStack.CODEC::write)?.requireNoNulls()?.foldRight(ListTag()) { tag, acum -> acum.also { it.add(tag) } }?.run { nbt.put("customDrops", this) }
    }

    fun canClaim(playerUUID: UUID): Boolean {
        if (lootMode == LootMode.UNLIMITED) return true

        val claim = this.getLootClaim(playerUUID)

        if (claim.isEmpty) {
            addClaimer(playerUUID)
            return true
        }
        if (lootMode.isTimeEnabled) if (claim.get().isExpired) {
            claim.get().refresh()
            return true
        }
        return false
    }

    fun getLootClaim(playerUUID: UUID): Optional<LootClaim> {
        return claims.stream().filter { claim: LootClaim -> claim.uuid == playerUUID }.findFirst()
    }

    fun addClaimer(playerUUID: UUID) {
        claims.add(LootClaim(playerUUID, Instant.now().plus(GenerationsCore.CONFIG.lootTime)))
    }


    fun removeClaimer(playerUUID: UUID) {
        claims.removeIf { a: LootClaim -> a.uuid == playerUUID }
    }

    fun shouldBreakBlock(): Boolean {
        return lootMode.isBreakable
    }

    val isCustomDrop: Boolean
        get() = this.doesCustomDrops && this.customDrops != null

    fun setCustomDrops(vararg customDrops: ItemStack?) {
        this.doesCustomDrops = true
        this.customDrops = NonNullList.of(ItemStack.EMPTY, *customDrops)
    }

    fun addCustomDrop(drop: ItemStack) {
        this.doesCustomDrops = true
        if (this.customDrops == null) this.customDrops = NonNullList.create()
        customDrops!!.add(drop)
    }

    enum class LootMode {
        ONCE_PER_PLAYER, TIMED, ONCE, UNLIMITED;

        val isDropOnce: Boolean
            get() = this == ONCE_PER_PLAYER || this == ONCE

        val isTimeEnabled: Boolean
            get() = this == TIMED

        val isBreakable: Boolean
            get() = this == ONCE
    }

    override fun getAngle(): Float {
        if (blockState.hasProperty(HorizontalDirectionalBlock.FACING)) {
            var dir = blockState.getValue(HorizontalDirectionalBlock.FACING)

            when (dir) {
                Direction.EAST -> dir = Direction.WEST
                Direction.WEST -> dir = Direction.EAST
                else -> {}
            }

            return dir.opposite.toYRot()
        }
        return 0f
    }
}

fun <T> T?.ifNotNull(function: (T) -> Unit) = this?.run(function)
